src/Microsoft.AspNetCore.Hosting/WebHostExtensions.cs

```diff
@@ -16,18 +19,32 @@ public static class WebHostExtensions
         /// <param name="host">The <see cref="IWebHost"/> to run.</param>
         public static void Run(this IWebHost host)
         {
+            var done = new ManualResetEvent(false);
             using (var cts = new CancellationTokenSource())
             {
-                Console.CancelKeyPress += (sender, eventArgs) =>
+                Action shutdown = () =>
                 {
-                    Console.WriteLine("Application is shutting down...");
-                    cts.Cancel();
+                    if (!cts.IsCancellationRequested)
+                    {
+                        Console.WriteLine("Application is shutting down...");
+                        cts.Cancel();
+                    }
 
+                    done.WaitOne();
+                };
+
+#if NETSTANDARD1_5
+                AssemblyLoadContext.Default.Unloading += context => shutdown();
```

Comment:
So 80% of the case will be using this in program.Main(), for the 20% case though you might want to use the load context associated with this assembly instance.

```C#
var alc = AssemblyLoadContext.GetLoadContext(typeof(WebHostExtensions).GetTypeInfo().Assembly);
alc.Unloading  += context => shutdown();
```